torch/csrc/lazy/core/ir.cpp

```diff
@@ -127,6 +131,20 @@ const Shape& Node::shape(size_t output_index) const {
   return shapes_.at(output_index);
 }
 
+// Add the shape generated by the shape_fn
+void Node::addShapeFromFunction(const std::function<Shape()>& shape_fn) {
+  shapes_.emplace_back(getShapeFromFunction(shape_fn));
+}
+
+// Add a shape generated by the shape_fn
+Shape Node::getShapeFromFunction(const std::function<Shape()>& shape_fn) {
+  if (hasBackend()) {
```

Comment:
is this just to make test cases pass when a backend isn't registered?

it seems like a solid approach, but might be cleaner if we just require a backend exists and collapse this function into the other (addShapeFromFunction).  I like that you already made the backend interface provide a default trivial implementation.

Reply:
> might be cleaner if we just require a backend exists

as in raise an error if it isn't?

> collapse this function into the other (addShapeFromFunction)

I considered this, but we still need both a `addShapeFromFunction` so that the IR nodes can add shapes after construction and we still need a `computeShape` for external access (such as in the test)

Reply:
> as in raise an error if it isn't?

well, i think all over LTC code we do things like getBackend()->some_method().  We should probably put the check inside getBackend() if it isn't actually there, but i wouldn't want everyone calling it to have to check.

> need a computeShape for external access (such as in the test)
why can't the test just call getBackend()->computeShape or whatever?  it seems unnecessary to have this also as a Node api.   (maybe it made more sense when it had a backend and non-backend part, but if all it does is call backend, why have it at all

Reply:
> why can't the test just call getBackend()->computeShape or whatever? 

because we still the node's hash value for the caching. I did originally consider removing it (which is why there was that lingering `#include` to the TS backend in the cache test) but upon that realization changed it back

Reply:
> We should probably put the check inside getBackend() if it isn't actually there, but i wouldn't want everyone calling it to have to check.

Oh, it is already there. https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/backend/backend_interface.cpp#L16
Though, I think I'll keep the extra check I made there for the slightly more clear error message.